sklearn/partial_dependence.py

```diff
@@ -0,0 +1,577 @@
+        given column.
+
+    Returns
+    -------
+    grid : ndarray, shape=(n_points, X.shape[1])
+        All data points on the grid. n_points is always ``<= grid_resolution **
+        X.shape[1]``.
+    Z: list of ndarray
+        The values with which the grid has been created. The ndarrays may be of
+        different shape: either (grid_resolution,) or (n_unique_values,).
+    """
+    try:
+        assert len(percentiles) == 2
+    except (AssertionError, TypeError):
+        raise ValueError('percentiles must be a sequence of 2 elements.')
+    if not all(0. <= x <= 1. for x in percentiles):
+        raise ValueError('percentiles values must be in [0, 1].')
+    if percentiles[0] >= percentiles[1]:
+        raise ValueError('percentiles[0] must be strictly less '
+                         'than percentiles[1].')
+
+    if grid_resolution <= 1:
+        raise ValueError('grid_resolution must be strictly greater than 1.')
+
+    values = []
+    for feature in range(X.shape[1]):
+        uniques = np.unique(X[:, feature])
+        if uniques.shape[0] < grid_resolution:
+            # feature has low resolution use unique vals
+            axis = uniques
+        else:
+            # create axis based on percentiles and grid resolution
+            emp_percentiles = mquantiles(X, prob=percentiles, axis=0)
+            if np.allclose(emp_percentiles[0, feature],
+                           emp_percentiles[1, feature]):
+                raise ValueError('percentiles are too close to each other, '
+                                 'unable to build the grid.')
+            axis = np.linspace(emp_percentiles[0, feature],
+                               emp_percentiles[1, feature],
+                               num=grid_resolution, endpoint=True)
+        values.append(axis)
+
+    return cartesian(values), values
+
+
+def _partial_dependence_recursion(est, grid, target_variables):
+
+    if est.init is not None:
+        warnings.warn(
+            'Using recursion method with a non-constant init predictor will '
+            'lead to incorrect partial dependence values.',
+            UserWarning
+        )
+
+    # grid needs to be DTYPE
+    grid = np.asarray(grid, dtype=DTYPE, order='C')
+
+    n_trees_per_stage = est.estimators_.shape[1]
+    n_estimators = est.estimators_.shape[0]
+    learning_rate = est.learning_rate
+    averaged_predictions = np.zeros((n_trees_per_stage, grid.shape[0]),
+                                    dtype=np.float64, order='C')
+    for stage in range(n_estimators):
+        for k in range(n_trees_per_stage):
+            tree = est.estimators_[stage, k].tree_
+            _partial_dependence_tree(tree, grid, target_variables,
+                                     learning_rate, averaged_predictions[k])
+
+    return averaged_predictions
+
+
+def _partial_dependence_brute(est, grid, target_variables, X):
+
+    averaged_predictions = []
+    for new_values in grid:
```

Comment:
would it be simpler to use an itertools.product here than to completely materialise the grid just to be iterated through?

Reply:
Might be simpler for just `_partial_dependence_brute` but maybe not for the whole file, because `_partial_dependence_recursion` still needs the actual grid. At least now we have a more generic code, but no strong opinion

Reply:
I suppose this is very minor as it can be changed later if we see fit